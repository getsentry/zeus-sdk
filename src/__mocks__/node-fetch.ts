// tslint:disable-next-line:no-default-export
export default jest.fn();
