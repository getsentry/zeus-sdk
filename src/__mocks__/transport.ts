export class Transport {
  public static instance: Transport;
  public postForm: jest.Mock;
  public request: jest.Mock;
  public requestJson: jest.Mock;

  public constructor() {
    Transport.instance = this;
    this.postForm = jest.fn();
    this.request = jest.fn();
    this.requestJson = jest.fn();
  }
}
