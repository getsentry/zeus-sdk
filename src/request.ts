import fetch, { RequestInit, Response } from 'node-fetch';

/**
 * Request options.
 *
 * As opposed to the default fetch options, this interface only permits headers
 * to be passed as object. This is needed as headers are preprocessed by the
 * request utility.
 */
export type RequestOptions = {
  [P in keyof RequestInit]: P extends 'headers'
    ? Record<string, string>
    : RequestInit[P]
};

export { Response };

/**
 * Parses an error message from the given response.
 *
 * First, this function tries to parse an error message from the response body.
 * If this does not work, it falls back to the HTTP status text.
 *
 * @param response A fetch Response object.
 * @returns A promise that resolves the error.
 */
async function parseError(response: Response): Promise<Error> {
  try {
    const { message } = (await response.json()) as { message?: string };
    return new Error(message || `${response.status} ${response.statusText}`);
  } catch (e) {
    return new Error(`${response.status} ${response.statusText}`);
  }
}

/**
 * Performs an AJAX request to the given url with the specified options using
 * fetch.
 *
 * After the request has finished, the result is parsed and checked for errors.
 * In case of an error, the response message is thrown as an error. On success,
 * the response object is passed into the promise.
 *
 * @param url The destination of the AJAX call.
 * @param options Options to the fetch call.
 * @returns A Promise to the parsed response body.
 */
export async function request(
  url: string,
  options: RequestOptions = {}
): Promise<Response> {
  const response = await fetch(url, options);
  if (!response.ok) {
    throw await parseError(response);
  }
  return response;
}
