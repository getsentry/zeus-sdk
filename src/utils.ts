import { URL } from 'whatwg-url';
import { Logger } from './transport';

/** Noop operation. */
export function noop(): void {
  // noop
}

export const noopLogger: Logger = {
  error: noop,
  info: noop,
  log: noop,
  warn: noop,
};

/**
 * Validates the given URL and makes sure it ends with a trailing slash.
 *
 * If the URL is not valid, an Error with details is thrown.
 *
 * @param url A fully qualified URL to sanitize.
 * @returns The sanitized URL.
 */
export function sanitizeURL(url: string): string {
  let sanitized = new URL(url).toString();

  if (!sanitized.endsWith('/')) {
    sanitized += '/';
  }

  return sanitized;
}

/**
 * Extracts ZEUS_HOOK_BASE URL from the environment or raises an error.
 *
 * @returns The Zeus hook base.
 */
export function getHookBase(): string {
  const base = process.env.ZEUS_HOOK_BASE;
  if (!base) {
    throw new Error('Missing ZEUS_HOOK_BASE environment variable');
  }
  return base;
}
