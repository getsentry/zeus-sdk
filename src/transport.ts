import FormData = require('form-data');
import { Response } from 'node-fetch';
import * as util from 'util';
import { URL } from 'whatwg-url';
import { request, RequestOptions } from './request';
import { noop, sanitizeURL } from './utils';

/** Default URL of the Zeus server. */
export const DEFAULT_URL = 'https://zeus.ci/';

/** A console logger. */
export interface Logger {
  /** Logs a message with default level. */
  log(...args: any[]): void;
  /** Logs an error message. */
  error(...args: any[]): void;
  /** Logs a warning message. */
  warn(...args: any[]): void;
  /** Logs an info message. */
  info(...args: any[]): void;
  /** Logs a debug message. */
  debug?(...args: any[]): void;
}

/** Options possed to the Zeus client constructor. */
export interface Options {
  /** The URL to the Zeus server. Defaults to https://zeus.ci. */
  url?: string;
  /** Authentication token for the API. */
  token?: string;
  /** A logger with the same interface like console. */
  logger?: Logger;
}

/** Options used in `Transport.requestJson`. */
export type JsonRequestOptions = {
  [P in keyof RequestOptions]: P extends 'body'
    ? Record<string, any>
    : RequestOptions[P]
};

/** Allowed value types for form fields. */
export type FormValue = string | number | Buffer;

/** A form field specification for `Transport.postForm`. */
export type FormField = FormValue | [FormValue, FormData.AppendOptions];

/** Form fields used in `Transport.postForm`. */
export interface FormFields {
  [key: string]: FormField | null | undefined;
}

/** Internal helper class used to send requests to Zeus. */
export class Transport {
  /** The URL to the Zeus server. Defaults to https://zeus.ci. */
  private readonly url: string;
  /** Authentication token for the API. */
  private readonly token: string | undefined;
  /** A logger with the same interface like console. */
  private readonly logger: Logger;

  /**
   * Creates a new API client.
   *
   * @param options Optional parameters for the client.
   */
  public constructor(options: Options = {}) {
    this.url = sanitizeURL(options.url || process.env.ZEUS_URL || DEFAULT_URL);
    this.token = options.token || process.env.ZEUS_TOKEN || undefined;
    this.logger = options.logger || {
      error: noop,
      info: noop,
      log: noop,
      warn: noop,
    };

    if (!this.logger.debug) {
      this.logger.debug = noop;
    }
  }

  /**
   * Computes the absolute URL of an endpoint specified by the given path.
   *
   * @param path Path to an endpoint.
   * @returns The absolute URL to the endpoint.
   */
  public getUrl(path: string = ''): string {
    return new URL(path, this.url).toString();
  }

  /** Returns the token used to authenticate with Zeus. */
  public getToken(): string | undefined {
    return this.token;
  }

  /**
   * Performs an API request to the given path.
   *
   * The request is performed relative to the configured url, unless the path is
   * a fully qualified URL with protocol and host. In that case, the request is
   * performed to the path as is.
   *
   * If configured, the authorization token is added to the request headers. It
   * can be overriden by passing a value for the "Authorization" header.
   *
   * On success, the raw result is passed into the promise
   *
   * @param path The endpoint of the API call.
   * @param options Options to the `fetch` call.
   * @returns A Promise to the parsed response body.
   */
  public async requestRaw(
    path: string,
    options: RequestOptions = {}
  ): Promise<Response> {
    const headers: Record<string, string> = { ...options.headers };
    const token = this.getToken();
    if (token !== undefined && !headers.Authorization) {
      headers.Authorization = `Bearer ${token.toLowerCase()}`;
    }

    const url = this.getUrl(path);
    return request(url, { ...options, headers }, this.logger);
  }

  /**
   * Performs an API request and converts the response to JSON.
   *
   * If configured, the authorization token is added to the request headers. It
   * can be overriden by passing a value for the "Authorization" header.
   *
   * On success, the parsed JSON result is passed into the promise. In case the
   * request fails or returns with a status code outside of the 200
   * range, an error is thrown with the response message field or status text.
   *
   * @param path The endpoint of the API call.
   * @param options Options to the `fetch` call.
   * @returns A Promise to the parsed response body.
   */
  public async request<T>(
    path: string,
    options: RequestOptions = {}
  ): Promise<T> {
    const headers: Record<string, string> = {
      Accept: 'application/json',
      ...options.headers,
    };

    const response = await this.requestRaw(path, { ...options, headers });
    return response.status === 204 ? undefined : response.json();
  }

  /**
   * Convert body to JSON and performs an API request to the given path.
   *
   * Null/undefined values in request body are stripped, and then the body
   * is serialized to JSON.
   *
   * @param path The endpoint of the API call.
   * @param options Options to the `fetch` call.
   * @returns A Promise to the parsed response body.
   */
  public async requestJson<T>(
    path: string,
    options: JsonRequestOptions = {}
  ): Promise<T> {
    const headers = {
      'Content-Type': 'application/json',
      ...options.headers,
    };

    // Clear null values from body
    const body: Record<string, any> = { ...options.body };
    Object.keys(body).forEach(key => {
      if (body[key] === null || body[key] === undefined) {
        delete body[key]; // tslint:disable-line:no-dynamic-delete
      }
    });

    return this.request<T>(path, {
      ...options,
      body: JSON.stringify(body),
      headers,
    });
  }

  /**
   * Posts the given form the the specified endpoint.
   *
   * This prevents requests with chunked transfer encoding by specifying
   * Content-Length of the body explicitly. This might lead to higher resource
   * consumption while sending the request, but works around servers that do not
   * support chunked requests.
   *
   * TODO: Once Zeus server has been updated to support chunked requests, this
   * method should be removed again.
   *
   * @param path The endpoint of the API call.
   * @param data Key value pairs to be posted to the server.
   * @returns The parsed server response.
   */
  public async postForm<T>(path: string, data: FormFields): Promise<T> {
    const form = new FormData();
    Object.keys(data).forEach(key => {
      const value = data[key];
      if (value !== undefined && value !== null) {
        // tslint:disable-next-line:deprecation
        if (util.isArray(value)) {
          form.append(key, value[0], value[1]);
        } else {
          form.append(key, value);
        }
      }
    });

    const length = await new Promise<number>((resolve, reject) => {
      form.getLength((e, len) => {
        if (e) {
          reject(e);
        } else {
          resolve(len);
        }
      });
    });

    const headers = {
      ...form.getHeaders(),
      'Content-Length': `${length}`,
    };

    return this.request<T>(path, {
      body: form,
      headers,
      method: 'POST',
    });
  }
}
