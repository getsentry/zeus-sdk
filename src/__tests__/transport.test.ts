import { Response } from 'node-fetch';
import { request } from '../request';
import { DEFAULT_URL, FormField, Transport } from '../transport';
import { noop, noopLogger } from '../utils';

jest.mock('../request');

const requestMock: jest.Mock<typeof request> = request as any;
const defaultTransportOptions = {
  logger: noopLogger,
};

describe('Transport', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    requestMock.mockReturnValue(
      Promise.resolve(new Response(JSON.stringify({ some: 'data' })))
    );

    // This causes some issues in Node 9.4.0 inside jest
    console.debug = noop;
  });

  describe('constructor', () => {
    const env = { ...process.env };

    afterEach(() => {
      process.env = { ...env };
    });

    test('initializes with the default URL', () => {
      const transport = new Transport(defaultTransportOptions);
      expect(transport.getUrl()).toBe(DEFAULT_URL);
    });

    test('reads from the ZEUS_URL environment variable', () => {
      const url = 'https://example.org/';
      process.env.ZEUS_URL = url;
      const transport = new Transport(defaultTransportOptions);
      expect(transport.getUrl()).toBe(url);
    });

    test('reads from the url option', () => {
      const url = 'https://example.org/';
      process.env.ZEUS_URL = 'invalid';
      const transport = new Transport({ ...defaultTransportOptions, url });
      expect(transport.getUrl()).toBe(url);
    });

    test('ensures a valid base URL', () => {
      const url = 'https://example.org/path';
      const transport = new Transport({ url });
      expect(transport.getUrl()).toBe(`${url}/`);
    });

    test('fails with an invalid URL', () => {
      const url = 'https://';
      expect(() => new Transport({ url })).toThrow(/valid URL/);
    });

    test('initializes with an empty token', () => {
      const transport = new Transport(defaultTransportOptions);
      expect(transport.getToken()).toBeUndefined();
    });

    test('reads from the ZEUS_TOKEN environment variable', () => {
      const token = 'zeus-u-1234567890';
      process.env.ZEUS_TOKEN = token;
      const transport = new Transport(defaultTransportOptions);
      expect(transport.getToken()).toBe(token);
    });

    test('reads from the token option', () => {
      const token = 'zeus-u-1234567890';
      process.env.ZEUS_TOKEN = 'invalid';
      const transport = new Transport({ ...defaultTransportOptions, token });
      expect(transport.getToken()).toBe(token);
    });
  });

  describe('request', () => {
    test('issues a request without options', async () => {
      const transport = new Transport(defaultTransportOptions);
      await transport.request('something');
      expect(requestMock).toMatchSnapshot();
    });

    test('resolves the response value', async () => {
      const transport = new Transport(defaultTransportOptions);
      const response = await transport.request('something');
      expect(response).toEqual({ some: 'data' });
    });

    test('accepts options', async () => {
      const options = { method: 'POST', body: '{"some":"data"}' };
      const transport = new Transport(defaultTransportOptions);
      await transport.request('something', options);
      expect(requestMock).toMatchSnapshot();
    });

    test('accepts absolute URLs', async () => {
      const transport = new Transport(defaultTransportOptions);
      await transport.request('http://example.org');
      expect(requestMock).toMatchSnapshot();
    });

    test('adds the Authorization token header', async () => {
      const headers = { Accept: 'application/json' };
      const token = 'zeus-u-1234567890';

      const transport = new Transport({ ...defaultTransportOptions, token });
      await transport.request('something', { headers });
      expect(requestMock).toMatchSnapshot();
    });

    test('accepts Authorization header overrides', async () => {
      const headers = { Authorization: 'custom' };
      const token = 'zeus-u-1234567890';

      const transport = new Transport({ ...defaultTransportOptions, token });
      await transport.request('something', { headers });
      expect(requestMock).toMatchSnapshot();
    });

    test('rejects invalid URLs', async () => {
      const transport = new Transport(defaultTransportOptions);
      try {
        await transport.request('///');
      } catch (e) {
        expect((e as Error).toString()).toMatch(/invalid url/i);
      }
    });

    test('resolves the parsed JSON result for status 200', async () => {
      const transport = new Transport(defaultTransportOptions);
      requestMock.mockImplementation(() => {
        const resp = new Response(JSON.stringify({ some: 'data' }));
        return Promise.resolve(resp);
      });

      const response = await transport.request('something');
      expect(response).toEqual({ some: 'data' });
    });

    test('resolves undefined for status 204', async () => {
      const transport = new Transport(defaultTransportOptions);

      requestMock.mockImplementation(
        async () =>
          new Response(JSON.stringify({ some: 'data' }), {
            status: 204,
          })
      );

      const response = await transport.request('something');
      expect(response).toBeUndefined();
    });
  });

  describe('requestJson', () => {
    let transport: Transport;

    beforeEach(() => {
      transport = new Transport(defaultTransportOptions);
    });

    test('serializes a JSON payload into the request', async () => {
      const options = {
        body: {
          a: [1, 2],
          b: 'c',
        },
        method: 'POST',
      };

      await transport.requestJson('something', options);
      expect(requestMock).toMatchSnapshot();
    });

    test('defaults to an empty body object', async () => {
      const options = {
        body: undefined,
        method: 'POST',
      };

      await transport.requestJson('something', options);
      expect(requestMock).toMatchSnapshot();
    });

    test('skips null and undefined keys', async () => {
      const options = {
        // tslint:disable-next-line:no-null-keyword
        body: { some: 'data', a: null, b: undefined },
        method: 'POST',
      };

      await transport.requestJson('something', options);
      expect(requestMock).toMatchSnapshot();
    });

    test('merges headers into the request', async () => {
      const options = {
        body: {
          a: [1, 2],
          b: 'c',
        },
        headers: {
          'X-Test': 'foo bar',
        },
        method: 'POST',
      };

      await transport.requestJson('something', options);
      expect(requestMock).toMatchSnapshot();
    });
  });

  describe('postForm', () => {
    let transport: Transport;

    beforeEach(() => {
      transport = new Transport(defaultTransportOptions);
    });

    test('calls request with a form data and correct headers', async () => {
      const data = { some: 'data' };
      await transport.postForm('something', data);
      expect(requestMock).toMatchSnapshot();
    });

    test('does not include null or undefined values', async () => {
      // tslint:disable-next-line:no-null-keyword
      const data = { some: 'data', a: null, b: undefined };
      await transport.postForm('something', data);
      expect(requestMock).toMatchSnapshot();
    });

    test('passes append options to FormData', async () => {
      const data = {
        some: ['data', { filename: 'other' }] as FormField,
      };

      await transport.postForm('something', data);
      expect(requestMock).toMatchSnapshot();
    });
  });
});
