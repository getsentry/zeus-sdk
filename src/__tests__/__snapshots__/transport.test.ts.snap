// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transport postForm calls request with a form data and correct headers 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://zeus.ci/something",
      Object {
        "body": FormData {
          "data": Object {
            "some": Object {
              "options": undefined,
              "value": "data",
            },
          },
        },
        "headers": Object {
          "Accept": "application/json",
          "Content-Length": "42",
          "content-type": "multipart/form-data; boundary=---feedface",
        },
        "method": "POST",
      },
      Object {
        "debug": [Function],
        "error": [Function],
        "info": [Function],
        "log": [Function],
        "warn": [Function],
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Promise {},
    },
  ],
}
`;

exports[`Transport postForm does not include null or undefined values 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://zeus.ci/something",
      Object {
        "body": FormData {
          "data": Object {
            "some": Object {
              "options": undefined,
              "value": "data",
            },
          },
        },
        "headers": Object {
          "Accept": "application/json",
          "Content-Length": "42",
          "content-type": "multipart/form-data; boundary=---feedface",
        },
        "method": "POST",
      },
      Object {
        "debug": [Function],
        "error": [Function],
        "info": [Function],
        "log": [Function],
        "warn": [Function],
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Promise {},
    },
  ],
}
`;

exports[`Transport postForm passes append options to FormData 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://zeus.ci/something",
      Object {
        "body": FormData {
          "data": Object {
            "some": Object {
              "options": Object {
                "filename": "other",
              },
              "value": "data",
            },
          },
        },
        "headers": Object {
          "Accept": "application/json",
          "Content-Length": "42",
          "content-type": "multipart/form-data; boundary=---feedface",
        },
        "method": "POST",
      },
      Object {
        "debug": [Function],
        "error": [Function],
        "info": [Function],
        "log": [Function],
        "warn": [Function],
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Promise {},
    },
  ],
}
`;

exports[`Transport request accepts Authorization header overrides 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://zeus.ci/something",
      Object {
        "headers": Object {
          "Accept": "application/json",
          "Authorization": "custom",
        },
      },
      BufferedConsole {
        "_buffer": Array [],
        "_counters": Object {},
        "_getSourceMaps": [Function],
        "_groupDepth": 0,
        "_timers": Object {},
        "assert": [Function],
        "clear": [Function],
        "count": [Function],
        "countReset": [Function],
        "debug": [Function],
        "dir": [Function],
        "error": [Function],
        "group": [Function],
        "groupCollapsed": [Function],
        "groupEnd": [Function],
        "info": [Function],
        "log": [Function],
        "time": [Function],
        "timeEnd": [Function],
        "trace": [Function],
        "warn": [Function],
        Symbol(counts): Map {},
        Symbol(groupIndent): "",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Promise {},
    },
  ],
}
`;

exports[`Transport request accepts absolute URLs 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "http://example.org/",
      Object {
        "headers": Object {
          "Accept": "application/json",
        },
      },
      Object {
        "debug": [Function],
        "error": [Function],
        "info": [Function],
        "log": [Function],
        "warn": [Function],
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Promise {},
    },
  ],
}
`;

exports[`Transport request accepts options 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://zeus.ci/something",
      Object {
        "body": "{\\"some\\":\\"data\\"}",
        "headers": Object {
          "Accept": "application/json",
        },
        "method": "POST",
      },
      Object {
        "debug": [Function],
        "error": [Function],
        "info": [Function],
        "log": [Function],
        "warn": [Function],
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Promise {},
    },
  ],
}
`;

exports[`Transport request adds the Authorization token header 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://zeus.ci/something",
      Object {
        "headers": Object {
          "Accept": "application/json",
          "Authorization": "Bearer zeus-u-1234567890",
        },
      },
      BufferedConsole {
        "_buffer": Array [],
        "_counters": Object {},
        "_getSourceMaps": [Function],
        "_groupDepth": 0,
        "_timers": Object {},
        "assert": [Function],
        "clear": [Function],
        "count": [Function],
        "countReset": [Function],
        "debug": [Function],
        "dir": [Function],
        "error": [Function],
        "group": [Function],
        "groupCollapsed": [Function],
        "groupEnd": [Function],
        "info": [Function],
        "log": [Function],
        "time": [Function],
        "timeEnd": [Function],
        "trace": [Function],
        "warn": [Function],
        Symbol(counts): Map {},
        Symbol(groupIndent): "",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Promise {},
    },
  ],
}
`;

exports[`Transport request issues a request without options 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://zeus.ci/something",
      Object {
        "headers": Object {
          "Accept": "application/json",
        },
      },
      Object {
        "debug": [Function],
        "error": [Function],
        "info": [Function],
        "log": [Function],
        "warn": [Function],
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Promise {},
    },
  ],
}
`;

exports[`Transport requestJson defaults to an empty body object 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://zeus.ci/something",
      Object {
        "body": "{}",
        "headers": Object {
          "Accept": "application/json",
          "Content-Type": "application/json",
        },
        "method": "POST",
      },
      Object {
        "debug": [Function],
        "error": [Function],
        "info": [Function],
        "log": [Function],
        "warn": [Function],
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Promise {},
    },
  ],
}
`;

exports[`Transport requestJson merges headers into the request 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://zeus.ci/something",
      Object {
        "body": "{\\"a\\":[1,2],\\"b\\":\\"c\\"}",
        "headers": Object {
          "Accept": "application/json",
          "Content-Type": "application/json",
          "X-Test": "foo bar",
        },
        "method": "POST",
      },
      Object {
        "debug": [Function],
        "error": [Function],
        "info": [Function],
        "log": [Function],
        "warn": [Function],
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Promise {},
    },
  ],
}
`;

exports[`Transport requestJson serializes a JSON payload into the request 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://zeus.ci/something",
      Object {
        "body": "{\\"a\\":[1,2],\\"b\\":\\"c\\"}",
        "headers": Object {
          "Accept": "application/json",
          "Content-Type": "application/json",
        },
        "method": "POST",
      },
      Object {
        "debug": [Function],
        "error": [Function],
        "info": [Function],
        "log": [Function],
        "warn": [Function],
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Promise {},
    },
  ],
}
`;

exports[`Transport requestJson skips null and undefined keys 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://zeus.ci/something",
      Object {
        "body": "{\\"some\\":\\"data\\"}",
        "headers": Object {
          "Accept": "application/json",
          "Content-Type": "application/json",
        },
        "method": "POST",
      },
      Object {
        "debug": [Function],
        "error": [Function],
        "info": [Function],
        "log": [Function],
        "warn": [Function],
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Promise {},
    },
  ],
}
`;
